server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -111,6 +98,16 @@ Q query() {
         return query;
     }
 
+    /**
+     * Is this an inefficient union of the top level query with the filter?
+     * If the top level query if complex we can't efficiently merge it with
+     * the filter. If we can't do that it is likely faster to just run the
+     * "native" aggregation implementation rather than go filter by filter.
+     */
+    public boolean isInefficientUnion() {
+        return false;
```

Comment:
The default is that filters are no "inefficient unions". In the case where we're fairly sure it *is* an inefficient union we override to true and bail as fast as we can.

Reply:
The trick here is that when we rewrite we're fairly sure we're making efficient queries. Well, we check for runtime fields queries, but otherwise, we expect rewriting to make *fairly* efficient queries.

When you are using the `filters` aggregator you very well could provide a terrible query. But it's going to be the same speed to run it filter by filter as it would be to run it in the compatible way. There are likely some cases where it'd have been faster to rewrite it. Or, better yet, to use the "disjunction collector" idea that I showed you to run all of the filters in parallel with the top level query. But this change should conservatively fall back to the old performance. Which wasn't great, but at least it wasn't terrible.